server/services/order.js

```diff
@@ -0,0 +1,82 @@
+module.exports = class OrderService {
+    static ORDER_BUILD_AND_SEND_CARRIER = "ORDER_BUILD_AND_SEND_CARRIER";
+
+    constructor(carrierService, waypointService, starService, starUpgradeService) {
+        this.carrierService = carrierService;
+        this.waypointService = waypointService;
+        this.starService = starService;
+        this.starUpgradeService = starUpgradeService;
+        this.orderHandlers = {
+            ORDER_BUILD_AND_SEND_CARRIER: this._handleBuildAndSendCarrier.bind(this)
+        }
+    }
+
+    async processOrdersForPlayer(game, player) {
+        if (!player.scheduledOrders) {
+            player.scheduledOrders = [];
+        }
+
+        const newOrders = [];
+
+        for (let order of player.scheduledOrders) {
+            let success = false;
+            try {
+                success = await this._performOrder(game, player, order.orderType, order.data);
+            } finally {
+                if (!success) {
+                    const retryPolicy = order.retryPolicy || "delete";
+                    if (retryPolicy === "retry") {
+                        newOrders.push(order);
+                    }
+                }
+            }
+        }
+
+        player.scheduledOrders = newOrders;
+    }
+
+    async _performOrder(game, player, orderType, data) {
+        const orderHandler = this.orderHandlers[orderType];
+        if (orderHandler) {
+            try {
+                return await orderHandler(game, player, data);
+            } catch (e) {
+                console.error(e);
+            }
+        }
+
+        return false;
+    }
+
+    async _handleBuildAndSendCarrier(game, player, data) {
+        if (!data.waypoints || !data.waypoints.length) {
+            return false;
+        }
+
+        const expenseConfig = game.constants.star.infrastructureExpenseMultipliers[game.settings.specialGalaxy.carrierCost];
+        const carrierCost = this.starUpgradeService.calculateCarrierCost(game, expenseConfig);
+
+        const startingStar = this.starService.getById(game, data.waypoints[0].source);
+        
+        const carriersAtStar = this.carrierService.getCarriersAtStar(game, startingStar._id);
+        const carriersOfPlayer = this.carrierService.listCarriersOwnedByPlayer(carriersAtStar, player._id);
+        // For simplicity, only handle carriers that have no specialists
+        const availableCarriers = carriersOfPlayer.filter(carrier => !carrier.specialistId && (!carrier.waypoints || !carrier.waypoints.length));
```

Comment:
Check for carriers **orbiting** the star isn't enough, we need to consider **established loops**.

Consider the following:
- **Tick 1:** 
- A loop is created between Star A and Star B
- **Tick 2:** 
- The carrier from the previous loop launches
- The AI logic sees that there is no carrier at the star and builds another one to set up another loop.

If there is already a carrier in orbit or in flight with the **same source and destination** then do not build another carrier.

Reply:
This is pretty much avoided because the logistics setup is only ran once, but maybe it could be improved by my suggestion above (storing the logistics graph in the db and checking if it needs to be changed)